File: SvgEditor.java
// SvgEditor.java
// Версия на Java с javax.xml.parsers, org.w3c.dom, Stream API

import javax.xml.parsers.*;
import javax.xml.transform.*;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;
import org.w3c.dom.*;
import java.io.*;
import java.util.*;
import java.util.stream.IntStream;

public class SvgEditor {
    private final String filename;
    private Document doc;
    private Element svgRoot;
    private List<Element> layers; // все группы верхнего уровня

    public SvgEditor(String filename) throws Exception {
        this.filename = filename;
        DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
        DocumentBuilder builder = factory.newDocumentBuilder();
        doc = builder.parse(new File(filename));
        svgRoot = doc.getDocumentElement();
        loadLayers();
    }

    private void loadLayers() {
        layers = new ArrayList<>();
        NodeList children = svgRoot.getChildNodes();
        for (int i = 0; i < children.getLength(); i++) {
            Node node = children.item(i);
            if (node.getNodeType() == Node.ELEMENT_NODE) {
                Element elem = (Element) node;
                // считаем группы или элементы с атрибутом data-layer
                if ("g".equals(elem.getTagName()) || elem.hasAttribute("data-layer")) {
                    layers.add(elem);
                }
            }
        }
    }

    public void listLayers() {
        if (layers.isEmpty()) {
            System.out.println("Слои не найдены.");
            return;
        }
        IntStream.range(0, layers.size()).forEach(i -> {
            Element e = layers.get(i);
            String id = e.getAttribute("id");
            if (id.isEmpty()) id = e.getAttribute("data-layer");
            System.out.printf("[%d] %s : %s%n", i, id.isEmpty() ? "без id" : id, attrsToString(e));
        });
    }

    private String attrsToString(Element e) {
        NamedNodeMap attrs = e.getAttributes();
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < attrs.getLength(); i++) {
            Attr a = (Attr) attrs.item(i);
            sb.append(a.getName()).append("=\"").append(a.getValue()).append("\" ");
        }
        return sb.toString();
    }

    public void addLayer(String id, Map<String, String> extraAttrs) {
        Element newGroup = doc.createElement("g");
        if (id != null && !id.isEmpty()) newGroup.setAttribute("id", id);
        extraAttrs.forEach(newGroup::setAttribute);
        svgRoot.appendChild(newGroup);
        layers.add(newGroup);
        System.out.println("Слой '" + id + "' добавлен.");
    }

    public void removeLayer(int index) {
        if (index < 0 || index >= layers.size()) throw new IndexOutOfBoundsException("Индекс вне диапазона");
        Element layer = layers.remove(index);
        svgRoot.removeChild(layer);
        System.out.println("Слой " + index + " удалён.");
    }

    public void moveLayer(int index, String direction) {
        if (index < 0 || index >= layers.size()) throw new IndexOutOfBoundsException("Индекс вне диапазона");
        int newIdx = direction.equals("up") ? index - 1 : index + 1;
        if (newIdx < 0 || newIdx >= layers.size()) throw new IllegalArgumentException("Невозможно переместить");
        // меняем местами в DOM
        Element el1 = layers.get(index);
        Element el2 = layers.get(newIdx);
        svgRoot.insertBefore(el1, el2);
        svgRoot.insertBefore(el2, el1); // не работает корректно, лучше использовать replaceChild? 
        // Проще: удалить и вставить заново.
        // Для простоты переставим в списке и пересоздадим дерево? Нет, проще сделать swap.
        // Перемещаем элементы через removeChild и insertBefore
        svgRoot.removeChild(el1);
        svgRoot.removeChild(el2);
        // Вставляем в правильном порядке
        if (index < newIdx) {
            svgRoot.insertBefore(el2, el1); // el1 уже удалён, но ещё в памяти
            svgRoot.insertBefore(el1, el2);
        } else {
            svgRoot.insertBefore(el1, el2);
            svgRoot.insertBefore(el2, el1);
        }
        // обновляем список
        Collections.swap(layers, index, newIdx);
        System.out.println("Слой " + index + " перемещён " + direction);
    }

    public void editLayer(int index, String attr, String value) {
        if (index < 0 || index >= layers.size()) throw new IndexOutOfBoundsException("Слой не найден");
        Element layer = layers.get(index);
        layer.setAttribute(attr, value);
        System.out.println("Атрибут '" + attr + "' слоя " + index + " установлен в '" + value + "'.");
    }

    public void save(String output) throws TransformerException {
        String out = (output == null || output.isEmpty()) ? filename : output;
        TransformerFactory factory = TransformerFactory.newInstance();
        Transformer transformer = factory.newTransformer();
        transformer.setOutputProperty(OutputKeys.ENCODING, "UTF-8");
        transformer.setOutputProperty(OutputKeys.INDENT, "yes");
        transformer.setOutputProperty("{http://xml.apache.org/xslt}indent-amount", "2");
        DOMSource source = new DOMSource(doc);
        StreamResult result = new StreamResult(new File(out));
        transformer.transform(source, result);
        System.out.println("Сохранено в " + out);
    }

    public static void main(String[] args) throws Exception {
        if (args.length < 2) {
            System.err.println("Использование: java SvgEditor <input.svg> <команда> [параметры]");
            System.exit(1);
        }
        String input = args[0];
        String command = args[1];
        SvgEditor editor = new SvgEditor(input);

        try {
            switch (command) {
                case "list":
                    editor.listLayers();
                    break;
                case "add-layer": {
                    String id = null;
                    Map<String, String> extra = new HashMap<>();
                    for (int i = 2; i < args.length; i++) {
                        if (args[i].equals("--id") && i+1 < args.length) {
                            id = args[++i];
                        } else if (args[i].equals("--attr") && i+1 < args.length) {
                            String kv = args[++i];
                            String[] parts = kv.split("=", 2);
                            if (parts.length == 2) extra.put(parts[0], parts[1]);
                        }
                    }
                    editor.addLayer(id, extra);
                    editor.save(null);
                    break;
                }
                case "remove-layer": {
                    int index = -1;
                    for (int i = 2; i < args.length; i++) {
                        if (args[i].equals("--index") && i+1 < args.length) {
                            index = Integer.parseInt(args[++i]);
                        }
                    }
                    if (index < 0) throw new IllegalArgumentException("Не указан --index");
                    editor.removeLayer(index);
                    editor.save(null);
                    break;
                }
                case "move-layer": {
                    int index = -1;
                    String dir = "";
                    for (int i = 2; i < args.length; i++) {
                        if (args[i].equals("--index") && i+1 < args.length) {
                            index = Integer.parseInt(args[++i]);
                        } else if (args[i].equals("--direction") && i+1 < args.length) {
                            dir = args[++i];
                        }
                    }
                    if (index < 0 || dir.isEmpty()) throw new IllegalArgumentException("Не указаны --index и --direction");
                    editor.moveLayer(index, dir);
                    editor.save(null);
                    break;
                }
                case "edit-layer": {
                    int index = -1;
                    String attr = "", value = "";
                    for (int i = 2; i < args.length; i++) {
                        if (args[i].equals("--index") && i+1 < args.length) {
                            index = Integer.parseInt(args[++i]);
                        } else if (args[i].equals("--attr") && i+1 < args.length) {
                            attr = args[++i];
                        } else if (args[i].equals("--value") && i+1 < args.length) {
                            value = args[++i];
                        }
                    }
                    if (index < 0 || attr.isEmpty() || value.isEmpty())
                        throw new IllegalArgumentException("Не указаны --index, --attr, --value");
                    editor.editLayer(index, attr, value);
                    editor.save(null);
                    break;
                }
                case "save": {
                    String output = null;
                    for (int i = 2; i < args.length; i++) {
                        if (args[i].equals("--output") && i+1 < args.length) {
                            output = args[++i];
                        }
                    }
                    editor.save(output);
                    break;
                }
                default:
                    System.err.println("Неизвестная команда: " + command);
            }
        } catch (Exception e) {
            System.err.println("Ошибка: " + e.getMessage());
            e.printStackTrace();
            System.exit(1);
        }
    }
}

File: svg_editor.rb
# svg_editor.rb
# Версия на Ruby с REXML, метапрограммированием

require 'rexml/document'
include REXML

class SvgEditor
  attr_reader :filename, :doc, :root, :layers

  def initialize(filename)
    @filename = filename
    @doc = Document.new(File.read(filename))
    @root = @doc.root
    load_layers
  end

  def load_layers
    @layers = []
    @root.each_element do |elem|
      if elem.name == 'g' || elem.attributes['data-layer']
        @layers << elem
      end
    end
  end

  def list_layers
    if @layers.empty?
      puts "Слои не найдены."
      return
    end
    @layers.each_with_index do |el, i|
      id = el.attributes['id'] || el.attributes['data-layer'] || 'без id'
      attrs = el.attributes.map { |k,v| "#{k}=\"#{v}\"" }.join(' ')
      puts "[#{i}] #{id} : #{attrs}"
    end
  end

  def add_layer(id, extra_attrs = {})
    new_group = Element.new('g')
    new_group.attributes['id'] = id if id
    extra_attrs.each { |k,v| new_group.attributes[k] = v }
    @root.add_element(new_group)
    @layers << new_group
    puts "Слой '#{id}' добавлен."
  end

  def remove_layer(index)
    if index < 0 || index >= @layers.size
      raise "Индекс вне диапазона"
    end
    el = @layers.delete_at(index)
    @root.delete_element(el)
    puts "Слой #{index} удалён."
  end

  def move_layer(index, direction)
    if index < 0 || index >= @layers.size
      raise "Индекс вне диапазона"
    end
    new_idx = direction == 'up' ? index - 1 : index + 1
    if new_idx < 0 || new_idx >= @layers.size
      raise "Невозможно переместить"
    end
    el1 = @layers[index]
    el2 = @layers[new_idx]
    # Перемещаем в DOM
    @root.delete_element(el1)
    @root.delete_element(el2)
    if index < new_idx
      @root.insert_before(el1, el2)
      @root.insert_before(el2, el1)
    else
      @root.insert_before(el2, el1)
      @root.insert_before(el1, el2)
    end
    @layers[index], @layers[new_idx] = @layers[new_idx], @layers[index]
    puts "Слой #{index} перемещён #{direction}."
  end

  def edit_layer(index, attr, value)
    if index < 0 || index >= @layers.size
      raise "Слой не найден"
    end
    @layers[index].attributes[attr] = value
    puts "Атрибут '#{attr}' слоя #{index} установлен в '#{value}'."
  end

  def save(output = nil)
    out = output || @filename
    File.write(out, @doc.to_s)
    puts "Сохранено в #{out}"
  end
end

# CLI
if ARGV.size < 2
  puts "Использование: ruby svg_editor.rb <input.svg> <команда> [параметры]"
  exit 1
end

input = ARGV[0]
command = ARGV[1]
editor = SvgEditor.new(input)

begin
  case command
  when 'list'
    editor.list_layers
  when 'add-layer'
    id = nil
    extra = {}
    i = 2
    while i < ARGV.size
      case ARGV[i]
      when '--id'
        id = ARGV[i+1]; i += 2
      when '--attr'
        kv = ARGV[i+1].split('=', 2)
        extra[kv[0]] = kv[1] if kv.size == 2
        i += 2
      else
        i += 1
      end
    end
    editor.add_layer(id, extra)
    editor.save
  when 'remove-layer'
    index = nil
    i = 2
    while i < ARGV.size
      if ARGV[i] == '--index'
        index = ARGV[i+1].to_i
        break
      end
      i += 1
    end
    raise "Не указан --index" if index.nil?
    editor.remove_layer(index)
    editor.save
  when 'move-layer'
    index = nil
    dir = nil
    i = 2
    while i < ARGV.size
      case ARGV[i]
      when '--index'
        index = ARGV[i+1].to_i; i += 2
      when '--direction'
        dir = ARGV[i+1]; i += 2
      else
        i += 1
      end
    end
    raise "Не указаны --index и --direction" if index.nil? || dir.nil?
    editor.move_layer(index, dir)
    editor.save
  when 'edit-layer'
    index = nil
    attr = nil
    value = nil
    i = 2
    while i < ARGV.size
      case ARGV[i]
      when '--index'
        index = ARGV[i+1].to_i; i += 2
      when '--attr'
        attr = ARGV[i+1]; i += 2
      when '--value'
        value = ARGV[i+1]; i += 2
      else
        i += 1
      end
    end
    raise "Не указаны --index, --attr, --value" if index.nil? || attr.nil? || value.nil?
    editor.edit_layer(index, attr, value)
    editor.save
  when 'save'
    output = nil
    i = 2
    while i < ARGV.size
      if ARGV[i] == '--output'
        output = ARGV[i+1]; break
      end
      i += 1
    end
    editor.save(output)
  else
    puts "Неизвестная команда: #{command}"
  end
rescue => e
  puts "Ошибка: #{e.message}"
  exit 1
end
